libraries/abstractions/ota_pal_psa/aws_ota_pal.c

```diff
@@ -0,0 +1,512 @@
+ *
+ * @param[in] C OTA file context information.
+ *
+ * @return The OTA PAL layer error code combined with the MCU specific error code. See OTA Agent
+ * error codes information in aws_iot_ota_agent.h.
+ *
+ * kOTA_Err_None is returned on success.
+ * kOTA_Err_SignatureCheckFailed is returned when cryptographic signature verification fails.
+ * kOTA_Err_BadSignerCert is returned for errors in the certificate itself.
+ * kOTA_Err_FileClose is returned when closing the file fails.
+ */
+OTA_Err_t prvPAL_CloseFile( OTA_FileContext_t * const C )
+{
+    /* Check the signature. */
+    if( prvPAL_CheckSignature( C ) != kOTA_Err_None)
+    {
+        return kOTA_Err_SignatureCheckFailed;
+    }
+
+    return kOTA_Err_None;
+}
+
+/**
+ * @brief Write a block of data to the specified file at the given offset.
+ *
+ * @note The input OTA_FileContext_t C is checked for NULL by the OTA agent before this
+ * function is called.
+ * The file pointer/handle C->pucFile, is checked for NULL by the OTA agent before this
+ * function is called.
+ * pacData is checked for NULL by the OTA agent before this function is called.
+ * ulBlockSize is validated for range by the OTA agent before this function is called.
+ * ulBlockIndex is validated by the OTA agent before this function is called.
+ *
+ * @param[in] C OTA file context information.
+ * @param[in] ulOffset Byte offset to write to from the beginning of the file.
+ * @param[in] pacData Pointer to the byte array of data to write.
+ * @param[in] ulBlockSize The number of bytes to write.
+ *
+ * @return The number of bytes written on a success, or a negative error code from the platform abstraction layer.
+ */
+int16_t prvPAL_WriteBlock( OTA_FileContext_t * const C,
+                           uint32_t ulOffset,
+                           uint8_t * const pcData,
+                           uint32_t ulBlockSize )
+{
+    OTA_PAL_PRINT( ( "\n\rprvPAL_WriteBlock: %d\n\r", ulBlockSize ) );
+    if( (C == NULL) || (C != sys_ctx ) || ( ota_image_id == 0 ) )
+    {
+        OTA_PAL_PRINT( ( "Invalid input argument!" ) );
+        return -1;
+    }
+
+    /* Call the TF-M Firmware Update service to write image data. */
+    if( psa_fwu_write( ( psa_image_id_t ) ota_image_id, ( size_t ) ulOffset , ( const void * )pcData, ( size_t ) ulBlockSize ) != PSA_SUCCESS)
+    {
+        return -3;
+    }
+
+    return ulBlockSize;
+}
+
+/**
+ * @brief Activate the newest MCU image received via OTA.
+ *
+ * This function shall do whatever is necessary to activate the newest MCU
+ * firmware received via OTA. It is typically just a reset of the device.
+ *
+ * @note This function SHOULD not return. If it does, the platform doesn't support
+ * an automatic reset or an error occurred.
+ *
+ * @return The OTA PAL layer error code combined with the MCU specific error code. See OTA Agent
+ * error codes information in aws_iot_ota_agent.h.
+ */
+OTA_Err_t prvPAL_ActivateNewImage( void )
+{
```

Comment:
The OTA process includes self-test phase , i.e when the the complete image is received its signature is verified and it is marked as in - Testing. The device then reboots and connects to AWS IoT services and updates the status after which image is accepted. If the device can not communicate back to the service then a timer resets the device and it should roll back to previous version. I do not see this support implemented.

Reply:
That is expected flow, thank you for sharing details.